From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sumit Gupta <sumitg@nvidia.com>
Date: Mon, 29 Jan 2024 19:39:50 +0530
Subject: [PATCH 243/254] NVIDIA: SAUCE: memory: tegra: init ICC once for
 multiple MC instances

Initialize the Memory Interconnect once if more than one MC instance
are present. All MC clients get added to the interconnect from the
clients table within Tegra MC driver and have unique ID. Client drivers
will use this ID (also present in their DT) to find the ICC node
irrespective of Socket. To find if the interconnect is already
initialized, use 'icc_node_create()' API which returns the existing
node with 'node->provider' if its already present.
Also, skip creating debugfs again if present to fix this error message.
  "debugfs: Directory 'mc' with parent '/' already present!"

Bug 4466730
Bug 4183070

Change-Id: Ib13e7bf346843cfd05eb0b55c002174f031f3b64
Signed-off-by: Sumit Gupta <sumitg@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/3065312
(cherry picked from commit 9d981375beb0dd6eeed16edf2a47e9a3bac39249)
Signed-off-by: Satish Seelamsetti <sseelamsetti@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/3108499
GVS: Gerrit_Virtual_Submit <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Bitan Biswas <bbiswas@nvidia.com>

diff --git a/drivers/memory/tegra/mc.c b/drivers/memory/tegra/mc.c
index 5fed2c54bee55..ad7420ff38784 100644
--- a/drivers/memory/tegra/mc.c
+++ b/drivers/memory/tegra/mc.c
@@ -863,6 +863,10 @@ static int tegra_mc_interconnect_setup(struct tegra_mc *mc)
 	if (IS_ERR(node))
 		return PTR_ERR(node);
 
+	/* Don't create duplicate entry if already created */
+	if (node->provider)
+		return 0;
+
 	node->name = "Memory Controller";
 	icc_node_add(node, &mc->provider);
 
@@ -949,8 +953,6 @@ static int tegra_mc_probe(struct platform_device *pdev)
 	if (IS_ERR(mc->regs))
 		return PTR_ERR(mc->regs);
 
-	mc->debugfs.root = debugfs_create_dir("mc", NULL);
-
 	if (mc->soc->ops && mc->soc->ops->probe) {
 		err = mc->soc->ops->probe(mc);
 		if (err < 0)
@@ -994,10 +996,13 @@ static int tegra_mc_probe(struct platform_device *pdev)
 			dev_err(&pdev->dev, "failed to register reset controller: %d\n", err);
 	}
 
-	err = tegra_mc_interconnect_setup(mc);
-	if (err < 0)
-		dev_err(&pdev->dev, "failed to initialize interconnect: %d\n",
-			err);
+	if (!debugfs_lookup("mc", NULL)) {
+		mc->debugfs.root = debugfs_create_dir("mc", NULL);
+
+		err = tegra_mc_interconnect_setup(mc);
+		if (err < 0)
+			dev_err(&pdev->dev, "failed to initialize interconnect: %d\n", err);
+	}
 
 	if (IS_ENABLED(CONFIG_TEGRA_IOMMU_SMMU) && mc->soc->smmu) {
 		mc->smmu = tegra_smmu_probe(&pdev->dev, mc->soc->smmu, mc);
diff --git a/drivers/memory/tegra/tegra186-emc.c b/drivers/memory/tegra/tegra186-emc.c
index 03ebab6fbe68f..873c9df072faf 100644
--- a/drivers/memory/tegra/tegra186-emc.c
+++ b/drivers/memory/tegra/tegra186-emc.c
@@ -277,6 +277,10 @@ static int tegra186_emc_interconnect_init(struct tegra186_emc *emc)
 	if (IS_ERR(node))
 		return PTR_ERR(node);
 
+	/* Don't create duplicate entry if already created */
+	if (node->provider)
+		return 0;
+
 	node->name = "External Memory Controller";
 	icc_node_add(node, &emc->provider);
 
-- 
2.43.0

